File: addon-sdk/source/test/test-environment.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

'use strict';

const { env } = require('sdk/system/environment');
const { Cc, Ci } = require('chrome');
const { get, set, exists } = Cc['@mozilla.org/process/environment;1'].
                             getService(Ci.nsIEnvironment);

exports['test exists'] = function(assert) {
  assert.equal('PATH' in env, exists('PATH'),
               'PATH environment variable is defined');
  assert.equal('FOO1' in env, exists('FOO1'),
               'FOO1 environment variable is not defined');
  set('FOO1', 'foo');
  assert.equal('FOO1' in env, true,
               'FOO1 environment variable was set');
  set('FOO1', null);
  assert.equal('FOO1' in env, false,
               'FOO1 environment variable was unset');
};

exports['test get'] = function(assert) {
  assert.equal(env.PATH, get('PATH'), 'PATH env variable matches');
  assert.equal(env.BAR2, undefined, 'BAR2 env variable is not defined');
  set('BAR2', 'bar');
  assert.equal(env.BAR2, 'bar', 'BAR2 env variable was set');
  set('BAR2', null);
  assert.equal(env.BAR2, undefined, 'BAR2 env variable was unset');
};

exports['test set'] = function(assert) {
  assert.equal(get('BAZ3'), '', 'BAZ3 env variable is not set');
  assert.equal(env.BAZ3, undefined, 'BAZ3 is not set');
  env.BAZ3 = 'baz';
  assert.equal(env.BAZ3, get('BAZ3'), 'BAZ3 env variable is set');
  assert.equal(get('BAZ3'), 'baz', 'BAZ3 env variable was set to "baz"');
};

exports['test unset'] = function(assert) {
  env.BLA4 = 'bla';
  assert.equal(env.BLA4, 'bla', 'BLA4 env varibale is set');
  delete env.BLA4;
  assert.equal(env.BLA4, undefined, 'BLA4 env variable is unset');
  assert.equal('BLA4' in env, false, 'BLA4 env variable no longer exists' );
};

require('test').run(exports);

File: python/mozperftest/mozperftest/tests/test_environment.py
#!/usr/bin/env python
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
from pathlib import Path
from unittest import mock

import mozunit
import pytest

from mozperftest.environment import MachEnvironment
from mozperftest.hooks import Hooks
from mozperftest.layers import Layer
from mozperftest.tests.support import get_running_env, requests_content

HERE = Path(__file__).parent.resolve()


def _get_env(hooks_path):
    return MachEnvironment(mock.MagicMock(), hooks=Hooks(mock.MagicMock(), hooks_path))


def test_run_hooks():
    env = _get_env(Path(HERE, "data", "hook.py"))
    assert env.hooks.run("doit", env) == "OK"


def test_bad_hooks():
    with pytest.raises(IOError):
        _get_env("Idontexists")


doit = [b"def doit(*args, **kw):\n", b"    return 'OK'\n"]


@mock.patch("mozperftest.utils.requests.get", requests_content(doit))
def test_run_hooks_url():
    env = _get_env("http://somewhere/hooks.py")
    assert env.hooks.run("doit", env) == "OK"


def test_layers():
    env = MachEnvironment(mock.MagicMock())
    assert env.get_layer("browsertime").name == "browsertime"


def test_context():
    mach, metadata, env = get_running_env()
    env.layers = [mock.MagicMock(), mock.MagicMock(), mock.MagicMock()]
    with env:
        env.run(metadata)


class FailureException(Exception):
    pass


class Failure(Layer):
    user_exception = True

    def run(self, metadata):
        raise FailureException()


def create_mock():
    m = mock.Mock()

    # need to manually set those
    def enter(self):
        self.setup()
        return self

    def exit(self, type, value, traceback):
        self.teardown()

    m.__enter__ = enter
    m.__exit__ = exit
    m.__call__ = mock.Mock()
    return m


def test_exception_return():
    # the last layer is not called, the error is swallowed
    hooks = str(Path(HERE, "data", "hook.py"))
    mach, metadata, env = get_running_env(hooks=hooks)
    last_layer = create_mock()
    env.layers = [create_mock(), Failure(env, mach), last_layer]
    with env:
        env.run(metadata)
    last_layer.assert_not_called()


def test_exception_resume():
    # the last layer is called, the error is swallowed
    hooks = str(Path(HERE, "data", "hook_resume.py"))
    mach, metadata, env = get_running_env(hooks=hooks)
    last_layer = create_mock()
    env.layers = [create_mock(), Failure(env, mach), last_layer]
    with env:
        env.run(metadata)
    last_layer.assert_called()


def test_exception_no_user_exception():
    # the last layer is called, the error is raised
    # because user_exception = False
    hooks = str(Path(HERE, "data", "hook_resume.py"))
    mach, metadata, env = get_running_env(hooks=hooks)
    last_layer = create_mock()
    f = Failure(env, mach)
    f.user_exception = False
    env.layers = [create_mock(), f, last_layer]
    with env, pytest.raises(FailureException):
        env.run(metadata)
    last_layer._call__.assert_not_called()


def test_exception_raised():
    # the error is raised
    hooks = str(Path(HERE, "data", "hook_raises.py"))
    mach, metadata, env = get_running_env(hooks=hooks)
    last_layer = create_mock()
    env.layers = [create_mock(), Failure(env, mach), last_layer]
    with env, pytest.raises(FailureException):
        env.run(metadata)
    last_layer.__call__.assert_not_called()


def test_metrics_last():
    mach, metadata, env = get_running_env()

    system = create_mock()
    browser = create_mock()

    # Check that the metrics layer is entered after
    # other have finished and that the other layers
    # were only called once
    class M:
        def __enter__(self):
            system.setup.assert_called_once()
            browser.setup.assert_called_once()
            system.teardown.assert_called_once()
            browser.teardown.assert_called_once()
            return self

        def __exit__(self, *args, **kw):
            return

        def __call__(self, metadata):
            return

    env.layers = [system, browser, M()]
    with env:
        env.run(metadata)


if __name__ == "__main__":
    mozunit.main()
